### PR TITLE
doc: Trainer.hyperparameter_search docstring discrepancy solved

### LINKED ISSUE #34239: `Trainer.hyperparameter_search` `kwargs` parameter has an inexact definition if using Optuna
According to the [docs](https://huggingface.co/docs/transformers/v4.45.2/en/main_classes/trainer#transformers.Trainer.hyperparameter_search.kwargs), those `kwargs` will only be used in the `create_study` method in Optuna.

However that is not exactly true, as we can see in `run_hp_search_optuna`. Some parameters from those `kwargs` will end up being used in the `optimize` method (which is great because it is what I wanted to do but the docs said I could not do it).

https://github.com/huggingface/transformers/blob/b54109c7466f6e680156fbd30fa929e2e222d730/src/transformers/integrations/integration_utils.py#L256

https://github.com/huggingface/transformers/blob/b54109c7466f6e680156fbd30fa929e2e222d730/src/transformers/integrations/integration_utils.py#L261

I didn't check ray or sigopt but I guess that something similar can be happening.

I don't really know what would be a good solution given that not every parameter in `optimize` is used and creating a `kwargs_optimize` would not make sense given that you support 2 other backends. Maybe a simple list of accepted paramters in the docstring?

If you tell me what solution would be the best, I could send a PR.

@stevhliu

### PR DESCRIPTION
**Pull request by GuillemGSubies**
# What does this PR do?

I tried to explain what `kwargs` do for all 3 backends.

Fixes #34239

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. #34239 
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@stevhliu

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -3428,13 +3428,18 @@ def hyperparameter_search(
...
                 A function that defines the trial/run name. Will default to None.
             kwargs (`Dict[str, Any]`, *optional*):
-                Additional keyword arguments passed along to `optuna.create_study` or `ray.tune.run`. For more
-                information see:
-
-                - the documentation of
-                  [optuna.create_study](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.study.create_study.html)
-                - the documentation of [tune.run](https://docs.ray.io/en/latest/tune/api_docs/execution.html#tune-run)
-                - the documentation of [sigopt](https://app.sigopt.com/docs/endpoints/experiments/create)
+                Additional keyword arguments passed along to the respective methods depending on the chosen backend:
```
Comment (stevhliu) [inline src/transformers/trainer.py:3431]:
```suggestion
                Additional keyword arguments for each backend:
```

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -3428,13 +3428,18 @@ def hyperparameter_search(
...
+
+                - If you use `optuna`, those will be the `kwargs` of
+                  [optuna.study.create_study](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.study.create_study.html)
+                  and also the parameters `timeout`, `n_jobs` and `gc_after_trial` correspondign to
+                  [optuna.study.Study.optimize](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.study.Study.html#optuna.study.Study.optimize)
```
Comment (stevhliu) [inline src/transformers/trainer.py:3436]:
```suggestion
                - `optuna`: parameters from
                  [optuna.study.create_study](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.study.create_study.html)
                  and also the parameters `timeout`, `n_jobs` and `gc_after_trial` from
                  [optuna.study.Study.optimize](https://optuna.readthedocs.io/en/stable/reference/generated/optuna.study.Study.html#optuna.study.Study.optimize)
```

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -3428,13 +3428,18 @@ def hyperparameter_search(
...
+                - If you use `ray`, those will be the `kwargs` of [tune.run](https://docs.ray.io/en/latest/tune/api_docs/execution.html#tune-run).
+                  If `resources_per_trial` is not set in the `kwargs` it will default to 1 CPU core and 1 GPU (if available).
+                  If `progress_reporter` is not set in the `kwargs`,
+                  [ray.tune.CLIReporter](https://docs.ray.io/en/latest/tune/api/doc/ray.tune.CLIReporter.html)will be used.
```
Comment (stevhliu) [inline src/transformers/trainer.py:3440]:
```suggestion
                - `ray`: parameters from [tune.run](https://docs.ray.io/en/latest/tune/api_docs/execution.html#tune-run).
                  If `resources_per_trial` is not set in the `kwargs`, it defaults to 1 CPU core and 1 GPU (if available).
                  If `progress_reporter` is not set in the `kwargs`,
                  [ray.tune.CLIReporter](https://docs.ray.io/en/latest/tune/api/doc/ray.tune.CLIReporter.html) is used.
```

File: src/transformers/trainer.py [hunk 4]
```diff
@@ -3428,13 +3428,18 @@ def hyperparameter_search(
...
+                - If you use `sigopt`, you can use the parameter `proxies` for
```
Comment (stevhliu) [inline src/transformers/trainer.py:3441]:
```suggestion
                - `sigopt`: the parameter `proxies` from
```

**Review by stevhliu — APPROVED** [review]
Thanks for clarifying the parameters! I added a few suggestions to make it less wordy

**Comment by GuillemGSubies** [comment]
Suggestions added, thank you 😊

**Comment by stevhliu** [comment]
Nice, if you can just rebase your branch on main to fix the failing `build_pr_documentation` test, that'd be great 🙂

**Comment by GuillemGSubies** [comment]
Ready

**Comment by stevhliu** [comment]
@LysandreJik the `run_tests` and `setup_and_quality` have been pending for quite some time now. Is there any way we can get these tests to run?

**Comment by stevhliu** [comment]
Can you maybe try closing this PR and opening a new one?

**Pull request closed without merging**